18.md

```diff
@@ -1,56 +1,96 @@
+The idea here is that the payer can identify itself when paying. The scope varies from free-form names (or [LUD-16](16.md) internet identifiers) useful for loose identification in comments or similar things, to cryptographic keys for proof of payment, to authentication keys for future login into a `SERVICE`'s website after a payment and other use cases.
 
-Payer MAY attach a `payerid` query parameter to LNURL-PAY callback with value set to JSON array:
+Aside from that, the payer ids are also committed to the invoice by means of the `descriptionHash` property, which ensures strong cryptographic guarantees for proof of payment.
+
+## 1. `payerIDs` record
+
+If `SERVICE` wants to get one or more types of payer identities from `WALLET` then it must alter its JSON response to the first callback to include a `payerIDs` field, as follows:
+
+```diff
+ {
+   callback: String,
+   maxSendable: MilliSatoshi,
+   minSendable: MilliSatoshi,
+   metadata: String,
++  payerIDs: {
++    "name": { "mandatory": Boolean },
++    "pubkey": { "mandatory": Boolean },
++    "identifier": { "mandatory": Boolean },
++    "email": { "mandatory": Boolean },
++    "auth": {
++       "mandatory": Boolean,
++       "k1": String // hex encoded 32 bytes of challenge
++    ],
++    ...other fields may be negotiated
++  },
+   tag: "payRequest",
+ }
+```
+
+## 2. Specifying payer identity before sending a payment
+
+In response to seeing a `payerIDs` record in initial response from `SERVICE`, `WALLET` attaches a `payerid` query parameter to LNURL-PAY callback with value set to a JSON object:
 
 ```diff
 - <callback><?|&>amount=<milliSatoshi>
-+ <callback><?|&>amount=<milliSatoshi>&payerid=[<identity record>, ...]
++ <callback><?|&>amount=<milliSatoshi>&payerid=<urlencode({json object})>
+```
+
+The JSON object MUST have the following format:
+
+```
+{
+  "name": <free form string>,
+  "pubkey": hex(<randomly generated secp256k1 pubkey>),
+  "auth": {
+    "key": hex(<linkingKey>),
+    "sig": hex(sign(utf8ToBytes(<k1>), <linkingPrivKey>)) // following LUD-04
+  },
+  "email": <email address>,
+  "identifier": <internet identifier>,
+  ...other fields may be included if supported by wallet and requested by service
+}
 ```
 
-JSON array may contain one or more records of the following forms:
+Each key in this JSON object should correspond to a requested payerid from the `payerIDs` record received from `SERVICE`.
 
-1. `[ "text/plain", <free-form string> ]`
-2. `[ "application/pubkey", hex(<randomly generated secp256k1 pubkey>) ]`
-3. `[ "application/lnurl-auth", hex(<linkingKey>), hex(sign(utf8ToBytes(<k1>), <linkingPrivKey>)) ]`
-4. `[ "text/identifier", <internet identifier> ]`
-5. `[ "text/email", <email address> ]`
+`WALLET` CAN send any of the payer id kinds if they are listed in the `payerIDs` record. But if any is marked as `"mandatory": true` `WALLET` must send it if it can, otherwise do not proceed with the payment flow.
```

Comment:
```diff
--- `WALLET` CAN send any of the payer id kinds if they are listed in the `payerIDs` record. But if any is marked as `"mandatory": true` `WALLET` must send it if it can, otherwise do not proceed with the payment flow.
+++ `WALLET` CAN send any of the payer id kinds if they are listed in the `payerIDs` record. But if any is marked as `"mandatory": true` `WALLET` MUST send it if it can, otherwise do not proceed with the payment flow.